### PR TITLE
Generation of DateTimes 4) Time based methods

### PR DESCRIPTION
**Pull request by zinki97**
See #175

In this step all time-based methods are implemented

### CODE REVIEW

**Comment by zinki97** [comment]
Have you changed something on the code style? @jlink

**Comment by jlink** [comment]
Maybe two weeks ago I aligned .editorconfig with the IntelliJ specific xml configuration. Any special thing you noticed?

**Comment by zinki97** [comment]
> Maybe two weeks ago I aligned .editorconfig with the IntelliJ specific xml configuration. Any special thing you noticed?

Only that about 400 lines were changed in `LocalDateTimeTests`.

**Comment by jlink** [comment]
> Only that about 400 lines were changed in `LocalDateTimeTests`.

which cleanly shows that the class is way too long.

**Comment by zinki97** [comment]
> > Only that about 400 lines were changed in `LocalDateTimeTests`.
> 
> which cleanly shows that the class is way too long.

What do you suggest? Packages with smaller classes?

**Comment by zinki97** [comment]
> > > Only that about 400 lines were changed in `LocalDateTimeTests`.
> > 
> > 
> > which cleanly shows that the class is way too long.
> 
> What do you suggest? Packages with smaller classes?

My suggest is to make smaller classes in their own packages. (At the end oft this step) For example package `LocalDateTimeTests` with class `ExhaustiveGeneration`, ...

What do you think @jlink ?

**Comment by jlink** [comment]
> My suggest is to make smaller classes in their own packages. (At the end oft this step) For example package `LocalDateTimeTests` with class `ExhaustiveGeneration`, ...
> 
> What do you think @jlink ?

Sounds reasonable. Pay attention to naming, not all names are picked up by JUnit by default. Stick with Java conventions: package names start lower case. 
What about „net.jqwik.time.localDateTime.ExhaustiveGenerationTests“

**Comment by zinki97** [comment]
> > My suggest is to make smaller classes in their own packages. (At the end oft this step) For example package `LocalDateTimeTests` with class `ExhaustiveGeneration`, ...
> > What do you think @jlink ?
> 
> Sounds reasonable. Pay attention to naming, not all names are picked up by JUnit by default. Stick with Java conventions: package names start lower case.
> What about „net.jqwik.time.localDateTime.ExhaustiveGenerationTests“

Thanks for the hint. Sounds good!

**Comment by zinki97** [comment]
`effectiveDate = date.plusDays(1);`

I think this line will cause an error if year/month/dayOfMonthBetween/etc. is set ://

**Comment by zinki97** [comment]
I also found a bug in `DefaultLocalTimeArbitrary`

**Comment by jlink** [comment]
> I also found a bug in `DefaultLocalTimeArbitrary`

More info? Failing test?

**Comment by zinki97** [comment]
> > I also found a bug in `DefaultLocalTimeArbitrary`
> 
> More info? Failing test?

The failing test was a new test in `LocalDateTimeTests`. After wondering why it wasn't working I found the problem in `DefaultLocalTimeArbitrary` after experimenting a bit. This shows me that there a soreme test cases missing.

Correct calculation of effective maximum time sometimes does not work if you set hours, minutes, seconds and a max time.

Example:
```
max-time: 13:25:31
hourBetween: 11, 11
minuteBetween: 22, 22
secondBetween: 31, 34
```

It calculates the max effective time `11:22:31`

**Comment by jlink** [comment]
> max-time: 13:25:31
> hourBetween: 11, 11
> minuteBetween: 22, 22
> secondBetween: 31, 34
> It calculates the max effective time `11:22:31` instead of `11:22:34`

That’s subtle.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java [hunk 1]
```diff
@@ -139,15 +141,24 @@ public static LocalTime calculateEffectiveMinWithPrecision(LocalTime effective,
...
+			effective = effective.withNano(999_999_999);
 			if (minuteBetween.getMax() < effective.getMinute()) {
 				effective = effective.withMinute(minuteBetween.getMax());
+				effective = effective.withSecond(59);
 				if (secondBetween.getMax() < effective.getSecond()) {
 					effective = effective.withSecond(secondBetween.getMax());
-					effective = effective.withNano(999_999_999);
 				}
 			}
 		}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:164]: I think I fixed the problem, but I will write more test cases.

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:164]: The problem is also present in calculating min values....

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java]: It should work now

**Comment by zinki97** [comment]
In the class `DefaultLocalDateArbitrary` filter prevents the bug. But in some cases edge cases are not generated.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java (hunk 1 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java]: There was still a bug, it has now been fixed

**Comment by zinki97** [comment]
> `effectiveDate = date.plusDays(1);`
> 
> I think this line will cause an error if year/month/dayOfMonthBetween/etc. is set ://

I thought about it: It calculates a wrong effective min date when month or dayOfMonth is restricted. But it doesn't matter. It don't affect exhaustive generation or edge cases. I'll add some tests to verify my thoughts.

I think a real min date is not necessary in this case. The LocalDateArbitrary class calculates the real min date in a later step.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java [hunk 2]
```diff
@@ -89,11 +112,19 @@ private int latestAllowedMonth() {
...
+		}
 		if (earliestMonth > effective.getMonth().getValue()) {
-			return effective.withMonth(earliestMonth);
+			effective = effective.withDayOfMonth(1).withMonth(earliestMonth);
 		}
 		if (dayOfMonthBetween.getMin() != null && dayOfMonthBetween.getMin() > effective.getDayOfMonth()) {
-			return effective.withDayOfMonth(dayOfMonthBetween.getMin());
+			try {
+				effective = effective.withDayOfMonth(dayOfMonthBetween.getMin());
+			} catch (DateTimeException e) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java:125]: Try-catch as case differentiation is a smell. Can you instead test for the condition that will lead to the exception?
If that is too much work, at least document why the exception can occur here.

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java]: Done

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java [hunk 3]
```diff
@@ -70,21 +70,74 @@
...
+						&& allowedMonths.get().contains(date.getMonth())
+				) {
+					return;
+				}
+			}
 		}
 		throw new IllegalArgumentException("These min/max configurations cannot be used together: No values are possible.");
 	}
 
+	private boolean valuesArePossible() {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java:92]: This looks terribly complicated. What happens if you just don't check? Won't the generator just fail at first attempt and tell that no value can be generated?

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java [hunk 4]
```diff
@@ -120,15 +121,28 @@ private LocalDate effectiveMinDate() {
 			effective = effective.withDayOfMonth(1).withMonth(earliestMonth);
 		}
 		if (dayOfMonthBetween.getMin() != null && dayOfMonthBetween.getMin() > effective.getDayOfMonth()) {
-			try {
+			if (isValidDate(effective.getYear(), effective.getMonth(), dayOfMonthBetween.getMin())) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateArbitrary.java:180]: I know I suggested to get rid of try-catch where appropriate. Here however you try (maybe successfully maybe not) to mimick what effective.withDayOfMonth() does in the library. I don't think that's well invested code.

My suggestion was to get rid of it OR DOCUMENT why an exception can occur here.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateTimeArbitrary.java [hunk 5]
```diff
@@ -101,15 +102,14 @@ private LocalDateTime calculateEffectiveMinDate(LocalDateTime effective) {
...
-			try {
+			if (date.getYear() < 999_999_999 || date.getMonth() != DECEMBER || date.getDayOfMonth() < 31) {
 				effectiveDate = date.plusDays(1);
-			} catch (DateTimeException dateTimeException) {
-				throw e;
+			} else {
+				throw new IllegalArgumentException("Cannot use this min value with precision " + ofPrecision.get());
 			}
 			date = effectiveDate;
 		}
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateTimeArbitrary.java:115]: Let's revisit this code together. IMO it does not communicate well what's happening here.

**Comment by jlink** [comment]
Can you roll back to https://github.com/jlink/jqwik/pull/184/commits/2c70dd4e23c2c9ff46235671f52e0ec07748a232? 
(And keep the two additional commits for a later PR).

Too much has happened. I'd like to get that into main first before proceeding with the new API methods.

**Comment by zinki97** [comment]
After our meeting it is clear what should happen with min value `11:52:29` and precision `MINUTES`. (exception)

But what about the max value `11:52:29` and precision `MINUTES`? Also exception or cut it to 11:52:00?

**Comment by jlink** [comment]
Yes. I think so. 

Von meinem iPhone gesendet

**Comment by zinki97** [comment]
> Yes. I think so.

What do you think? Exception? Or Cut it? :D

**Comment by jlink** [comment]
Exception. Gives us the option to later add an optional rounding parameter. 

Von meinem iPad gesendet

**Comment by jlink** [comment]
> Exception. Gives us the option to later add an optional rounding parameter. Von meinem iPad gesendet

Here's how it's being handled in big decimal generation:

```java
	@Property
	void test2(@ForAll("bigs") BigDecimal bigInteger) {

	}

	@Provide
	Arbitrary<BigDecimal> bigs() {
		return Arbitraries.bigDecimals()
						  .between(new BigDecimal("1.111"), new BigDecimal("2.222"))
						  .ofScale(2);
	}
```

Will throw the following exception on run:

```
net.jqwik.api.JqwikException:
    Decimal value 1.111 cannot be represented with scale 2.
    You may want to use a higher scale
```

I suggest a similiar revealing error message. I'm not quite sure if IllegalArgumentException would be preferrable.

**Comment by zinki97** [comment]
> I suggest a similiar revealing error message. I'm not quite sure if IllegalArgumentException would be preferrable.

The current exception is:

```
java.lang.IllegalArgumentException:
    Can't use 13:59:59.999999999 as maximum time with precision Seconds.
```

Should I change it?

**Comment by jlink** [comment]
> Should I change it?

You could add „You may want to round the time to Seconds or change precision to Nanos“

**Comment by zinki97** [comment]
> You may want to round the time to Seconds or change precision to Nanos

I've changed it to `You may want to round the time to Seconds or change the precision.`

**Comment by jlink** [comment]
> I've changed it to `You may want to round the time to Seconds or change the precision.`

Did you change it or did you add the sentence?

**Comment by zinki97** [comment]
> > I've changed it to `You may want to round the time to Seconds or change the precision.`
> 
> Did you change it or did you add the sentence?

Add the sentence
(next PR)

**Comment by jlink** [comment]
Build warning:
```
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/DateRange.java:17: warning - Tag @see: reference not found: LeapYears
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/DayOfMonthRange.java:17: warning - Tag @see: reference not found: LeapYears
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/DayOfWeekRange.java:18: warning - Tag @see: reference not found: LeapYears
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/MonthDayRange.java:18: warning - Tag @see: reference not found: LeapYears
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/MonthRange.java:19: warning - Tag @see: reference not found: LeapYears
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/YearMonthRange.java:18: warning - Tag @see: reference not found: LeapYears
/Users/link/Documents/git-projects/jqwik/time/src/main/java/net/jqwik/time/api/constraints/YearRange.java:18: warning - Tag @see: reference not found: LeapYears
```

**Comment by zinki97** [comment]
Could you please review this PR?

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java [hunk 6]
```diff
@@ -51,91 +44,69 @@
...
+				throwException = (effective.getNano() % 1_000) != 0;
+		}
+		if (throwException) {
+			throwDurationAndPrecisionException(effective.toString(), minimum, ofPrecision.get());
+		}
+	}
+
+	private static void throwDurationAndPrecisionException(String val, boolean minimum, ChronoUnit precision) {
+		String minMax = minimum ? "minimum" : "maximum";
+		throw new IllegalArgumentException("Can't use " + val + " as " + minMax + " duration with precision " + precision + ".\nYou may want to round the duration to " + precision + " or change the precision.");
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java:87]: Use String.format() instead of String concatenation.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java [hunk 7]
```diff
@@ -51,91 +44,69 @@
...
+	private Duration calculateEffectiveMax(DurationBetween durationBetween, OfPrecision ofPrecision) {
+		Duration effective = durationBetween.getMax() != null ? durationBetween.getMax() : calculateMaxPossibleValue(ofPrecision);
+		checkValueAndPrecision(effective, ofPrecision, false);
+		return effective;
+	}
+
+	private Duration calculateMaxPossibleValue(OfPrecision ofPrecision) {
+		switch (ofPrecision.get()) {
+			case HOURS:
+				return Duration.ofSeconds((Long.MAX_VALUE / (60 * 60)) * (60 * 60), 0);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java:99]: All those calculations could be on OfPrecision objects

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java [hunk 8]
```diff
@@ -51,91 +44,69 @@
...
-				}
-			}
-			return effective;
-		} catch (ArithmeticException e) {
-			throw new IllegalArgumentException("Max value must be decreased because of precision " + precision + " but results in a " + e.getMessage());
+	private void checkValueAndPrecision(Duration effective, OfPrecision ofPrecision, boolean minimum) {
+		boolean throwException = false;
+		switch (ofPrecision.get()) {
+			case HOURS:
+				throwException = effective.getSeconds() % 3600 != 0;
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java:68]: All those calculations could be in OfPrecision

**Review by jlink — COMMENTED** [review]
Some recommendations for class OfPrecision:
- make it a real immutable value object, get rid of those clone().set/setProgrammatically()
- it can do many calculations itself e.g. toSeconds(), compare(ChronoUnit) etc.
   Just look for how you use instances of it. It will help you get rid of the many switches in the using code.

**Comment by zinki97** [comment]
> * make it a real immutable value object, get rid of those clone().set/setProgrammatically()
> * it can do many calculations itself e.g. toSeconds(), compare(ChronoUnit) etc.
>   Just look for how you use instances of it. It will help you get rid of the many switches in the using code.

I don't really understand what you mean, sorry. Could you give me an example or explain it, please?

**Comment by jlink** [comment]
> I don't really understand what you mean, sorry. Could you give me an example or explain it, please?

Look at LocalDateTime for example. You can create an instance of it, but you cannot change its internal state. Comparing two instances will compare on their value. And LocalDateTime come with many methods for calculation, comparing, transformation etc.  Basically all classes in java.time follow that pattern. If OfPrecision would too, my feeling is that quite some code can become more elegant.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java [hunk 9]
```diff
@@ -200,7 +200,7 @@ public DurationArbitrary between(Duration min, Duration max) {
 	@Override
 	public DurationArbitrary ofPrecision(ChronoUnit ofPrecision) {
 		DefaultDurationArbitrary clone = typedClone();
-		clone.ofPrecision.set(ofPrecision);
+		clone.ofPrecision = this.ofPrecision.set(ofPrecision);
 		return clone;
 	}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java:206]: Is this what you mean?

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java [hunk 10]
```diff
@@ -180,7 +180,7 @@ private void setOfPrecisionImplicitly(DefaultDurationArbitrary clone, Duration d
 		}
 		ChronoUnit ofPrecision = DefaultLocalTimeArbitrary.calculateOfPrecisionFromNanos(duration.getNano());
 		if (clone.ofPrecision.get().compareTo(ofPrecision) > 0) {
-			clone.ofPrecision.setProgrammatically(ofPrecision);
+			clone.ofPrecision = this.ofPrecision.setProgrammatically(ofPrecision);
 		}
 	}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java:185]: Is this what you mean?

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java [hunk 11]
```diff
@@ -11,19 +11,34 @@
...
+	private OfPrecision set(ChronoUnit precision, boolean precisionSet) {
 		if (!ALLOWED_PRECISIONS.contains(precision)) {
 			throw new IllegalArgumentException("Precision value must be one of these ChronoUnit values: HOURS, MINUTES, SECONDS, MILLIS, MICROS, NANOS");
 		}
-		this.precision = precision;
+		if (this.precision.equals(precision) && this.precisionSet == precisionSet) {
+			return this;
+		}
+		return new OfPrecision(precision, precisionSet);
 	}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:42]: Is this what you mean?

**Review by zinki97 — COMMENTED** [review]
Is this what you mean? (some examples)

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java (hunk 11 above)
Comment (jlink) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:42]: It’s indeed going in the direction I meant, but seeing it it’s unsymmetrical to the Between objects. :-/
So, for symmetry reasons I‘d go back to the previous approach. 

Bringing the logic into OfPrecision, however, should still be an improvement. 

All that forth and back is an implication of asynchronous pull requests. If we were pair-programming, a lot of design ideas could be tried out and would prove themselves worthwhile or unworthy immediately.

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:42]: Isn't it better to change the other classes, too?

**Comment by jlink** [comment]
To me they feel like state holders not like state itself, but I'm unsure.
Which way will make the code simpler in your opinion?

Am Do., 29. Apr. 2021 um 19:06 Uhr schrieb zinki97 ***@***.***

**Comment by zinki97** [comment]
> To me they feel like state holders not like state itself, but I'm unsure. Which way will make the code simpler in your opinion?  .

The previous version is less code, the actual version of `OfPrecision` is more like `LocalDateTime`

**Comment by jlink** [comment]
I vorw for previous version in this case. Until we have new insights.

Am Do., 29. Apr. 2021 um 20:45 Uhr schrieb zinki97 ***@***.***

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java [hunk 12]
```diff
@@ -34,4 +35,57 @@ public boolean isSet() {
...
+			case MILLIS:
+				return LocalTime.of(23, 59, 59, 999_000_000);
+			case MICROS:
+				return LocalTime.of(23, 59, 59, 999_999_000);
+			case NANOS:
+				return LocalTime.of(23, 59, 59, 999_999_999);
+			default:
+				return LocalTime.of(23, 59, 59, 0);
+		}
+	}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:89]: I don't know if it is possible to make it simpler without switch, but I think not.

Is the class now like you want or what should I do next?

Comment (jlink) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:89]: Looks good to me.

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:89]: Should I change it in all classes?

Example: `calculateEffectiveMin()`

Comment (jlink) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/valueRanges/OfPrecision.java:89]: Unless you see a reason why calculate* is the better name in some cases

**Comment by zinki97** [comment]
Ready for review

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java [hunk 13]
```diff
@@ -18,126 +18,73 @@
...
-		} catch (ArithmeticException e) {
-			throw new IllegalArgumentException("Min value must be increased because of precision " + precision + " but results in a " + e.getMessage());
+	private Duration effectiveMin(DurationBetween durationBetween, OfPrecision ofPrecision) {
+		Duration effective = durationBetween.getMin() != null ? durationBetween.getMin() : minPossibleValue(ofPrecision);
+		checkValueAndPrecision(effective, ofPrecision, true);
+		return effective;
+	}
+
+	private Duration minPossibleValue(OfPrecision ofPrecision) {
+		switch (ofPrecision.get()) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultDurationArbitrary.java:54]: This calculation should also move to OfPrecision

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java [hunk 14]
```diff
@@ -51,7 +48,7 @@
 
 	}
 
-	static private LocalTime calculateLocalTime(long longAdd, LocalTime effectiveMin, OfPrecision precision) {
+	static private LocalTime localTimeFromValue(long longAdd, LocalTime effectiveMin, OfPrecision precision) {
 		switch (precision.get()) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:52]: move calculation to OfPrecision

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java [hunk 15]
```diff
@@ -68,7 +65,7 @@ static private LocalTime calculateLocalTime(long longAdd, LocalTime effectiveMin
 		}
 	}
 
-	private long calculateLongEnd(LocalTime effectiveMin, LocalTime effectiveMax) {
+	private long longsBetween(LocalTime effectiveMin, LocalTime effectiveMax) {
 		switch (ofPrecision.get()) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:69]: Move calculation to OfPrecision

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java [hunk 16]
```diff
@@ -85,97 +82,113 @@ private long calculateLongEnd(LocalTime effectiveMin, LocalTime effectiveMax) {
...
 
-	public static LocalTime calculateEffectiveMaxWithPrecision(LocalTime effective, OfPrecision ofPrecision) {
-		return calculateEffectiveMaxWithPrecision(effective, ofPrecision.get());
-	}
-
-	public static LocalTime calculateEffectiveMaxWithPrecision(LocalTime effective, ChronoUnit ofPrecision) {
-		switch (ofPrecision) {
-			case HOURS:
-				effective = effective.withMinute(0);
+	private static LocalTime setEffectiveMaxValues(LocalTime effective, OfPrecision ofPrecision, ChronoUnit precision) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:167]: Does the method name fit? Is there something being set?

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java [hunk 17]
```diff
@@ -85,97 +82,113 @@ private long calculateLongEnd(LocalTime effectiveMin, LocalTime effectiveMax) {
...
+				effective = ofPrecision.get().compareTo(MINUTES) <= 0 ? effective.withMinute(59) : effective.withMinute(0);
 			case SECONDS:
-				effective = effective.withNano(effective.getNano() % 1_000_000);
-			case MILLIS:
-				effective = effective.withNano((effective.getNano() / 1_000_000) * 1_000_000 + effective.getNano() % 1_000);
-			case MICROS:
-				effective = effective.withNano(effective.getNano() - effective.getNano() % 1_000);
+				effective = ofPrecision.get().compareTo(SECONDS) <= 0 ? effective.withSecond(59) : effective.withSecond(0);
+			default:
+				switch (ofPrecision.get()) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:174]: Move calculation to OfPrecision

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java [hunk 18]
```diff
@@ -85,97 +82,113 @@ private long calculateLongEnd(LocalTime effectiveMin, LocalTime effectiveMax) {
...
 			case MINUTES:
-				effective = effective.withSecond(0);
+				effective = ofPrecision.get().compareTo(MINUTES) <= 0 ? effective.withMinute(59) : effective.withMinute(0);
 			case SECONDS:
-				effective = effective.withNano(effective.getNano() % 1_000_000);
-			case MILLIS:
-				effective = effective.withNano((effective.getNano() / 1_000_000) * 1_000_000 + effective.getNano() % 1_000);
-			case MICROS:
-				effective = effective.withNano(effective.getNano() - effective.getNano() % 1_000);
+				effective = ofPrecision.get().compareTo(SECONDS) <= 0 ? effective.withSecond(59) : effective.withSecond(0);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java:172]: Can those comparisons be replaced by methods in OfPrecision? e.g. ofPrecision.isLessOrEqualTo(..)?

**Review by jlink — COMMENTED** [review]
I still see potential in giving OfPrecision more functionality that depends on it.
I marked some of those places but certainly not all.
Fix what you see. I'll do the rest after merging the PR. Well, maybe, some day.

**Review by jlink — APPROVED** [review]
I think we're done here :-)
